File: C++/Context.cpp
#include "Context.h"
#include "CustomException.h"

Context::Context():level(0)
{
}


Context* Context::GetInstance()
{
	static Context con;
	return &con;
}

const string& Context::Varification(char ch) const
{
	int temp_level = level;
	while (temp_level)
	{
		if (check[temp_level].name_var == ch)
			return check[temp_level].val_var;
		temp_level--;
	}
	throw CustomException(2);
}

void Context::SetVarible(char ch, const string& num)
{
	if (level > check.size())
		check.resize(2 * level);
	check[level].name_var = ch;
	check[level].val_var = num;
}

void Context::addbracklevel()
{
	level++;
}

void Context::subbracklevel()
{
	level--;
}

File: Python/context.py
import custom_exception


class Context:
    __level = [0]
    __check = [('0', '0')]

    def verification(self, character):
        temp_level = self.__level[0]
        while temp_level:
            name_var, val_var = self.__check[temp_level]
            if name_var == character:
                return val_var
            temp_level -= 1
        raise custom_exception.CustomException(2)

    def set_variable(self, character, string):
        self.__check.insert(self.__level[0], (character, string))

    def add_bracket_level(self):
        self.__level[0] += 1

    def sub_bracket_level(self):
        self.__level[0] -= 1
